### PR TITLE
ENH: Speed up sparse.csgraph.dijkstra

### PR DESCRIPTION
**Pull request by windows-server-2003**
`sparse.csgraph.dijkstra` implements the dijkstra algorithm to find the shortest path of a graph.
The comment in the code insists that the total time complexity of running it for all source vertices is $O(N(Nk + N\log(N)))$ where $N$ is the number of nodes and $k$ is the average number of connected edges per node.
However, it turned out that the worst case complexity is being $\Theta(N^3)$ in the current implementation. Star graph(graph with edges connecting node #0 and node #i for all 1 <= i < n) is one of the worst cases.
This is due to an incomplete implementation of Fibonacci heap.
<details>
<summary>Details of the implementation flaw</summary>
The number of children of a node in Fibonacci heap can be temporarily very large after repeated calls of `insert_node` because `insert_node` does nothing other than just appending a new child.<br>
`remove_min` iterates over these large number of children but it does not impact the overall complexity because `remove_min` rearranges those nodes so that there is no more such problematic situation around the nodes. In order for `remove_min` to iterate over a large number of children again, a large number of `insert_node` is needed, so amortized analysis works.  
However, this also means that we should not touch children when we do not resolve the high-rank vertex.<br>
In the current implementation, leftmost_sibling and rightmost_sibling(which both iterate over the children) are called in other context, causing the time complexity to bump.<br>
</details>
<details>
<summary>Details of the solution</summary>
Currently, node.children points to an arbitrary child of the node. In the new implementation, it always points to the leftmost child of the children. In other words, node.children.left_sibling is always NULL.<br>
Also, heap.min_node is also kept at the leftmost end of the linked list of roots.<br>
This way, there is no need to iterate over the children of a node in places other than remove_min.
</details>
The debugging function at the bottom of the file is also changed so that it is compatible with the new implementation(and with Python 3 !).<br>
Star graphs are added to the benchmark file, and a typo in the file name is fixed.

Here is the benchmark result of before/after on i5-10600k, VirtualBox Ubuntu 20.04. The new implementation is substantially faster with format = 'star' and slightly faster with format = 'random' which is the existing benchmark.
```
Before:
[100.00%] ··· ===== =============== ============= ================ ==============
              --                          min_only / format
              ----- -------------------------------------------------------------
                n    True / random   True / star   False / random   False / star
              ===== =============== ============= ================ ==============
                30      178±2μs        143±1μs       193±0.9μs       152±0.5μs
               300      417±5μs        204±1μs      5.05±0.02ms     1.80±0.01ms
               900    2.16±0.01ms      700±3μs       112±0.3ms       50.2±0.2ms
              ===== =============== ============= ================ ==============
After:
[100.00%] ··· ===== =============== ============= ================ ==============
              --                          min_only / format
              ----- -------------------------------------------------------------
                n    True / random   True / star   False / random   False / star
              ===== =============== ============= ================ ==============
                30      176±2μs       144±0.3μs      191±0.9μs       147±0.6μs
               300      380±2μs        163±1μs      4.08±0.03ms       522±3μs
               900    1.99±0.03ms      201±1μs       95.9±0.5ms     3.50±0.03ms
              ===== =============== ============= ================ ==============
```

### CODE REVIEW

**Comment by windows-server-2003** [comment]
I'm not sure if this belongs to MAINT or BUG.
Let me know if you have any idea on this.

**Comment by rgommers** [comment]
> I'm not sure if this belongs to MAINT or BUG.

I'd say ENH for significant performance improvements. This looks quite good, thanks @windows-server-2003

**Comment by windows-server-2003** [comment]
Thank you.
I'd like to rebase/force-push to edit the commit messages from MAINT to ENH and squash the "Comply with lint" one.

**Comment by rgommers** [comment]
@jjerphan would you be interested in having a look at this?

**Comment by jjerphan** [comment]
@rgommers: the "curious me" definitely wants to jump in it, yet the "busy me" is "Wait, look at you todo-max-heap, Julien". :smile: 

I might start on my free-time, probably in a few weeks?

I have a few PRs I also have started reviewing for SciPy, and it's probably better than I work toward having them merge first, IMO.

**Comment by rgommers** [comment]
That makes perfect sense, thanks!

**Comment by windows-server-2003** [comment]
@jjerphan I know you are busy, but look at the 14x speedup in the most fundamental function in graph area!  It will surely improve scipy a lot!

**Comment by jjerphan** [comment]
Hi @windows-server-2003,

I appreciate the speed-up brought by your PR. I have it on my list of reviews, unfortunately I am currently busy with other things at the moment. Thank you for your patience.

(Also, I like your GitHub handle.)

**Comment by windows-server-2003** [comment]
Okay, always thanks for working on scipy and related things!

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 1]
```diff
@@ -1497,11 +1489,14 @@ cdef void add_sibling(FibonacciNode* node, FibonacciNode* new_sibling):
...
-    new_sibling.left_sibling = temp
-    new_sibling.right_sibling = NULL
+    
+    # insert new_sibling between node and node.right_sibling
+    if node.right_sibling:
+        node.right_sibling.left_sibling = new_sibling
+    new_sibling.right_sibling = node.right_sibling
+    new_sibling.left_sibling = node
     new_sibling.parent = node.parent
+    node.right_sibling = new_sibling
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1499]: Nit.

```suggestion
    # Insert new_sibling between node and node.right_sibling
    if node.right_sibling:
        node.right_sibling.left_sibling = new_sibling
    new_sibling.right_sibling = node.right_sibling
    new_sibling.left_sibling = node
    node.right_sibling = new_sibling
    
    new_sibling.parent = node.parent
```

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 2]
```diff
@@ -1536,7 +1527,7 @@ ctypedef FibonacciNode* pFibonacciNode
 
 
 cdef struct FibonacciHeap:
-    FibonacciNode* min_node
+    FibonacciNode* min_node # min_node.left_sibling is always NULL
     pFibonacciNode[100] roots_by_rank  # maximum number of nodes is ~2^100.
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1531]: The number of roots (equals to `100` here and in other place) might better be [defined as a literal value via a Cython `DEF` statement](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html?highlight=%22DEF%22#compile-time-definitions):

```cython
DEF N_ROOTS = 100
```

and replace the hard-coded value.

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 3]
```diff
@@ -1546,9 +1537,14 @@ cdef void insert_node(FibonacciHeap* heap,
     #              - node is a valid pointer
     #              - node is not the child or sibling of another node
     if heap.min_node:
-        add_sibling(heap.min_node, node)
         if node.val < heap.min_node.val:
+            # heap.min_node is always at the leftmost end of the linked-list
+            node.left_sibling = NULL
+            node.right_sibling = heap.min_node
+            heap.min_node.left_sibling = node
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1544]:
```suggestion
            # Replace node with heap.min_node, which is always 
            # at the leftmost end of the roots' linked-list.
            node.left_sibling = NULL
            node.right_sibling = heap.min_node
            heap.min_node.left_sibling = node
```

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 4]
```diff
@@ -1566,6 +1562,9 @@ cdef void decrease_val(FibonacciHeap* heap,
         remove(node)
         insert_node(heap, node)
     elif heap.min_node.val > node.val:
+        remove(node)
+        node.right_sibling = heap.min_node
+        heap.min_node.left_sibling = node.right_sibling
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1567]:
```suggestion
        # Replace node with heap.min_node, which is always 
        # at the leftmost end of the roots' linked-list.
        remove(node)
        node.right_sibling = heap.min_node
        heap.min_node.left_sibling = node.right_sibling
```

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 5]
```diff
@@ -1510,12 +1505,8 @@ cdef void remove(FibonacciNode* node):
     # Assumptions: - node is a valid pointer
     if node.parent:
         node.parent.rank -= 1
-        if node.left_sibling:
-            node.parent.children = node.left_sibling
-        elif node.right_sibling:
+        if node.parent.children == node:
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1508]: I think `children` better be rename `leftmost_child` for the new representation.
```suggestion
        if node.parent.children == node:  # node is the leftmost sibling.
```

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 6]
```diff
@@ -1536,7 +1527,7 @@ ctypedef FibonacciNode* pFibonacciNode
 
 
 cdef struct FibonacciHeap:
-    FibonacciNode* min_node
+    FibonacciNode* min_node # min_node.left_sibling is always NULL
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1530]:
```suggestion
    # In this representation, min_node is always at the leftmost end
    # of the linked-list, hence min_node.left_sibling is always NULL.
    FibonacciNode* min_node
```

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 7]
```diff
@@ -1604,32 +1603,22 @@ cdef FibonacciNode* remove_min(FibonacciHeap* heap):
         unsigned int i
 
     # make all min_node children into root nodes
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1605]: If this constructs Binomial Trees, I think it's relevant indicating it (I can't suggest on this line unfortunately).

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 8]
```diff
@@ -1604,32 +1603,22 @@ cdef FibonacciNode* remove_min(FibonacciHeap* heap):
...
 
-    # remove min_node, and point heap to the new min
+    # remove min_root and choose another root as a preliminary min_root
     out = heap.min_node
+    temp = heap.min_node.right_sibling
     remove(heap.min_node)
     heap.min_node = temp
+    
+    if temp == NULL:
+        return out
```
Comment (jjerphan) [inline scipy/sparse/csgraph/_shortest_path.pyx:1621]: This pruning can be motivated.
```suggestion
    if temp == NULL:
        # There is a unique root in the tree, hence a unique node
        # which is the minimum that we return here.
        return out
```

Also, I think `out` can be rename to `min_root` or `min_node` (having a sole `min_` name for this variable and the heap attribute will ease reading IMO).

**Review by jjerphan — COMMENTED** [review]
Thank you for improving this implementation, @windows-server-2003!

Here are a few remark w.r.t to the algorithmic changes only.

I think the original implementation can be improved (mainly w.r.t to documentation or Cython technical details) as part of this PR or in subsequent ones.

In particular, I think that:
 - some function can be qualified as `inline` and `nogil`
 - the various step of the implementations can better be document, especially regarding re-linking and Binomial Tree construction (which I believe are used now?)
 - some comments can be turned into docstrings 
 - academic references to Fibonacci Heap can be given.
 - tests can be added to assert proper behavior and invariants (e.g. the min root being the roots' leftmost sibling.
 - the implementation of `Fibonacci{Node,Heap}` and associated function might be moved in a dedicated file.

File: scipy/sparse/csgraph/_shortest_path.pyx (hunk 1 above)
Comment (windows-server-2003) [reply scipy/sparse/csgraph/_shortest_path.pyx:1499]: I agree

File: scipy/sparse/csgraph/_shortest_path.pyx (hunk 2 above)
Comment (windows-server-2003) [reply scipy/sparse/csgraph/_shortest_path.pyx:1531]: Thanks, I'll replace it.

File: scipy/sparse/csgraph/_shortest_path.pyx (hunk 3 above)
Comment (windows-server-2003) [reply scipy/sparse/csgraph/_shortest_path.pyx:1544]: Yes, but did you mean "Replace heap.min_node with node"?

Comment (jjerphan) [reply scipy/sparse/csgraph/_shortest_path.pyx:1544]: Ah yes, my bad. I just have corrected it.

**Comment by windows-server-2003** [comment]
I just remembered that C++ STL can be used in Cython
std::priority_queue should replace the Fibonacci Heap, and

 - It reduces the maintenance cost
 - priority_queue doesn't have decrease_val but we can just push another node instead
 - I suspect priority_queue is faster(needs benchmark though)

If it's okay to use C++ STL in scipy, I want to test it

**Comment by windows-server-2003** [comment]
Here's the results of the benchmark(3 runs each, I'm only able to use the one-time benchmark `dev.py bench` because I haven't found the proper way of using libcpp from Cython yet)

<details>
<summary>Benchmark results</summary>

```
FibonacciHeap
[ 50.00%] ··· ===== =============== ============= ================ ==============
              --                          min_only / format
              ----- -------------------------------------------------------------
                n    True / random   True / star   False / random   False / star
              ===== =============== ============= ================ ==============
                30      387±0μs        496±0μs        390±0μs         513±0μs
               300      618±0μs        517±0μs        4.19±0ms        946±0μs
               900      2.48±0ms       553±0μs        90.2±0ms        4.38±0ms
              ===== =============== ============= ================ ==============
              ===== =============== ============= ================ ==============
                30      382±0μs        2.08±0ms       1.26±0ms        843±0μs
               300      1.04±0ms       896±0μs        7.31±0ms        946±0μs
               900      2.53±0ms       560±0μs        90.6±0ms        4.42±0ms
              ===== =============== ============= ================ ==============
              ===== =============== ============= ================ ==============
                30      411±0μs        492±0μs        400±0μs         505±0μs
               300      658±0μs        1.15±0ms       8.66±0ms        933±0μs
               900      2.57±0ms       560±0μs        90.0±0ms        4.42±0ms
              ===== =============== ============= ================ ==============

std::priority_queue
[ 50.00%] ··· ===== =============== ============= ================ ==============
              --                          min_only / format
              ----- -------------------------------------------------------------
                n    True / random   True / star   False / random   False / star
              ===== =============== ============= ================ ==============
                30      413±0μs        1.56±0ms       786±0μs         1.07±0ms
               300      1.18±0ms       657±0μs        3.06±0ms        1.21±0ms
               900      2.17±0ms       574±0μs        41.9±0ms        5.21±0ms
              ===== =============== ============= ================ ==============
              ===== =============== ============= ================ ==============
                30      400±0μs        2.18±0ms       1.76±0ms        854±0μs
               300      1.04±0ms       890±0μs        8.38±0ms        1.58±0ms
               900      2.21±0ms       600±0μs        42.0±0ms        5.15±0ms
              ===== =============== ============= ================ ==============
              ===== =============== ============= ================ ==============
                30      453±0μs        503±0μs        428±0μs         1.36±0ms
               300      1.54±0ms       570±0μs        2.51±0ms        2.82±0ms
               900      2.30±0ms       596±0μs        42.0±0ms        5.20±0ms
              ===== =============== ============= ================ ==============
```

</details>

Neither of the two constantly outperforms the other, but I think std::priority_queue is a better choice because

 - It shows significant(50%+) improvement in one of the tests while Fibonacci Heap is only 20% faster in the best case(among these testcases)
 - It would reduce 300 lines of code and gives better maintenability

What do you think?

**Comment by jjerphan** [comment]
Thanks for performing the comparison! 
By default `std::priority_queue` is backed by a `std::vector` seen as a binary heap. This option simpler and more efficient in practice than Fibonacci Heap when the number of elements is small to medium due to data locality mainly.

IIRC, the theoretical better complexity of Fibonacci Heaps really arises for use-case with large number of samples.

Hence, I would rerun the benchmark for `n in [10_000, 100_000, 1_000_000]` (if possible).

If `std::priority_queue` is competitive, I think it makes sense to replace Fibonacci Heap by `std::priority_queue` for eased maintenance. But it's up to SciPy's maintainers to decide.

---

As for using `std::priority_heap` from Cython, you should be able to `cimport` symbols via [the declaration set by Cython](https://github.com/cython/cython/blob/efdf996dced075682d86b7e53a12b56b5ebfea38/Cython/Includes/libcpp/queue.pxd):

```cython
from libcpp.priority_queue cimport queue, priority_queue
```

and use it similarly to as you would in C++, see [this test in the Cython code-base](https://github.com/cython/cython/blob/c25c87d71107e634162302f7f61a119eff539a48/tests/run/cpp_stl_cpp11.pyx#L61-L73).

You will also need to add support for C++ when creating the extension, that is you need to modify:

https://github.com/scipy/scipy/blob/06da9ee65cdee4a8aa036767489bfc5a030527f4/scipy/sparse/csgraph/setup.py#L10-L12

into:

```python
    config.add_extension('_shortest_path',
         sources=['_shortest_path.c'],
         include_dirs=[numpy.get_include()],
         language="c++",
)

**Comment by jjerphan** [comment]
Thank you for the new results, @windows-server-2003.

I guess another PR can be created to replace Fibonacci Heaps by `std::priority_queue`s, maintaining this one open in the meantime.

**Comment by tylerjereddy** [comment]
As far as I can tell, this PR should provide a performance improvement on its own. I read through all the comments above, and I don't really see any blockers, just suggestions for even more improvements.

Let me see if I can apply some of the suggestions, rebase (this hasn't flushed through CI since July), and push + run benchmarks locally.

**Comment by tylerjereddy** [comment]
The one CI failure I see so far is gh-17475, which is allowed because unrelated. Let's see how the rest of the CI and my local benchmarks proceed.

**Comment by tylerjereddy** [comment]
Rebased local benchmark results are solid with `asv`:

```
       before           after         ratio
     [b265d919]       [4328431c]
     <main>           <faster_dijkstra>
-       159±0.4ms        143±0.3ms     0.90  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(900, False, 'random')
-       571±0.9μs        501±0.7μs     0.88  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(300, True, 'random')
-     3.50±0.01ms      3.05±0.02ms     0.87  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(900, True, 'random')
-        7.28±0ms      6.06±0.01ms     0.83  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(300, False, 'random')
-       258±0.5μs        207±0.9μs     0.80  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(300, True, 'star')
-     2.37±0.01ms         717±10μs     0.30  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(300, False, 'star')
-         872±2μs        255±0.2μs     0.29  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(900, True, 'star')
-     64.1±0.08ms      5.18±0.06ms     0.08  sparse_csgraph_dijkstra.Dijkstra.time_dijkstra_multi(900, False, 'star')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

```

**Comment by tylerjereddy** [comment]
So, the CI is looking good, the performance improvement is good, the code review is positive and the suggestions I could easily apply have been applied. I note that the codecov line coverage report is not helpful here--it points to unrelated files. I'd have liked to see that, but not going to block the merge on that.

The separate commits follow our usual commit-writing practices so I'm happy to leave them as-is I think.

Thanks for your first contribution to SciPy.

**Comment by jjerphan** [comment]
#17019 might also be interesting integrating. IMHO, it is even better than this PR.

**Pull request merged**